From 33b3c128e0f480f6e421761b85c5f00289b62449 Mon Sep 17 00:00:00 2001
From: Kyle Hsiao <kylehsiao@google.com>
Date: Wed, 1 Oct 2025 11:57:54 +0000
Subject: [PATCH] [NFC] Fix use-after-free in eventCallback

Transform access to the static shared pointer 'Nfc::mCallback' to be thread-safe to prevent a use-after-free crash.

The use-after-free occurred when an asynchronous thread (eventCallback) attempted to call a method on 'mCallback' immediately after the main thread Nfc::open had destroyed the underlying object. The simple null check was insufficient due to the race condition.

This fix implements the correct shared pointer synchronization pattern:
1. Protects the read/write of 'mCallback' using 'mCallbackLock'.
2. Creates a local 'std::shared_ptr<INfcClientCallback> localCallback' inside the critical section. This local copy holds a temporary strong reference, guaranteeing the object's lifetime for the duration of the subsequent sendEvent() call.

Bug: 392699284
Test: nfc_service_fuzzer
Test: atest NfcNciUnitTests
Test: atest CtsNfcTestCases
Test: atest VtsAidlHalNfcTargetTest
Test: atest NfcTestCases
Test: atest NfcServiceTest
Change-Id: I8e9e83be7f939bbfc183cb3879808ceacf931450
(cherry picked from commit fc619c3348188ff0020cdeb7d4c4728a0246fe1a)
---
 aidl/Nfc.h | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/aidl/Nfc.h b/aidl/Nfc.h
index 707bd65..fd23e24 100644
--- a/aidl/Nfc.h
+++ b/aidl/Nfc.h
@@ -32,6 +32,8 @@ using ::aidl::android::hardware::nfc::NfcConfig;
 using ::aidl::android::hardware::nfc::NfcEvent;
 using ::aidl::android::hardware::nfc::NfcStatus;
 
+static pthread_mutex_t sCallbackLock = PTHREAD_MUTEX_INITIALIZER;
+
 // Default implementation that reports no support NFC.
 struct Nfc : public BnNfc {
  public:
@@ -52,7 +54,11 @@ struct Nfc : public BnNfc {
   binder_status_t dump(int fd, const char** args, uint32_t numArgs) override;
 
   static void eventCallback(uint8_t event, uint8_t status) {
-    if (mCallback != nullptr) {
+    std::shared_ptr<INfcClientCallback> localCallback;
+    pthread_mutex_lock(&sCallbackLock);
+    localCallback = mCallback;
+    pthread_mutex_unlock(&sCallbackLock);
+    if (localCallback != nullptr) {
       NfcEvent mEvent;
       NfcStatus mStatus;
       switch (event) {
@@ -96,7 +102,7 @@ struct Nfc : public BnNfc {
         default:
           mStatus = NfcStatus::FAILED;
       }
-      auto ret = mCallback->sendEvent(mEvent, mStatus);
+      auto ret = localCallback->sendEvent(mEvent, mStatus);
       if (!ret.isOk()) {
         LOG(ERROR) << "Failed to send event!";
       }
@@ -104,9 +110,13 @@ struct Nfc : public BnNfc {
   }
 
   static void dataCallback(uint16_t data_len, uint8_t* p_data) {
+    std::shared_ptr<INfcClientCallback> localCallback;
+    pthread_mutex_lock(&sCallbackLock);
+    localCallback = mCallback;
+    pthread_mutex_unlock(&sCallbackLock);
     std::vector<uint8_t> data(p_data, p_data + data_len);
-    if (mCallback != nullptr) {
-      auto ret = mCallback->sendData(data);
+    if (localCallback != nullptr) {
+      auto ret = localCallback->sendData(data);
       if (!ret.isOk()) {
         LOG(ERROR) << "Failed to send data!";
       }
-- 
2.53.0.335.g19a08e0c02-goog

